Fhir Patient Processing Integration

dixonwhitmire opened the pull request:
PR to address FHIR Patient Processing Issues 

Fixes #3510

- Added FhirServiceBase to define required FHIR processing operations - data mapping, query processing, etc
- Updated FhirPatientService to support FhirServiceBase expectations
- Implemented Fhir Patient query parameters per spec based on current OpenEMR support
- Updated tests and adjusted api/controller code and tests as necessary
- Updated PatientService to support new UUID registry and “standard” OpenEMR API processing.

Note: FHIR API Responses have not been updated to adhere to the FHIR R4 specification. Updating the API to support FHIR R4 is larger GSOC effort. From what I could tell the existing implementation wasn't uniform or complete. This PR just "unbreaks" the patient FHIR and OpenEMR API endpoints and integrates them with the "standard" OpenEMR API response.

sjpadgett commented on src/Services/FHIR/FhirPatientService.php:
```diff
@@ -2,14 +2,26 @@
 
-class FhirPatientService
+/**
+ * FHIR Patient Service Tests
+ * @coversDefaultClass OpenEMR\Services\FHIR\FhirPatientService
+ * @package   OpenEMR
+ * @link      http://www.open-emr.org
+ * @author    Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2020 Jerry Padgett <sjpadgett@gmail.com>
```
This doesn't seem right? Also don't forget to include yourself in authorship. This helps us know who to blame if something goes wrong!
Again, kidding:)

dixonwhitmire replied:
Ha! Will do.

sjpadgett commented on src/Services/FHIR/iFhirService.php:
```diff
@@ -0,0 +1,32 @@
+ * @copyright Copyright (c) 2020 Dixon Whitmire <dixonwh@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+interface iFhirService
```
Using interface classes in PHP always give me pause. Wouldn't an abstract class be more versatile? Up to you.
@yashrajbothra , you need to be aware here.

dixonwhitmire replied:
I'll update this to an abstract class in the next push. Yep for the moment we don't need to define any implementation details or defaults, but agreed that we should just go ahead and make this an abstract class, so that we have the freedom to do so.

sjpadgett commented on src/Services/FHIR/FhirPatientService.php:
```diff
@@ -32,25 +33,33 @@ public function __construct()
 
         $patientResource->setActive(true);
 
+        $narrativeText = '';
+        if (isset($data['fname'])) {
+            $narrativeText = $data['fname'];
+        }
+        if (isset($data['lname'])) {
+            $narrativeText .= ' ' . $data['lname'];
+        }
+        $text = array(
+            'status' => 'generated',
+            'div' => '<div xmlns="http://www.w3.org/1999/xhtml"> <p>' . $narrativeText . '</p></div>'
+        );
+        $patientResource->setText($text);
```
Happy to see this.

yashrajbothra replied:
BTW what should be the future of narration? Should we add narration in every resource or Patient Resource is special?

sjpadgett replied:
I'd say all Yash..

dixonwhitmire replied:
Here are some links to the "resource" hierarchy which may be helpful in determining which "base FHIR fields" should be included.  Each of the resources within OpenEMR's implementation are extensions of the [Domain Resource](https://www.hl7.org/fhir/domainresource.html) which in turn extends the [Base Resource](https://www.hl7.org/fhir/resource.html).

dixonwhitmire commented:
@bradymiller @sjpadgett @yashrajbothra - Hi Folks - I'm still working on implementing the FhirPatientService "insert" and "query" methods with the corresponding tests. Once that's done I'll be able to update the FhirPatientRestController, which shouldn't require much additional work, as the heavy lifting is in the service. In the meantime, comments/feedback are certainly welcome. I'll be sure to remove the [WIP] from the PR once it's ready for "review review" 😄 

At this point I haven't implemented the "getOne" or "update" methods as these both require associating a FHIR Resource ID with, in my case, the OpenEMR patient record. I didn't want to make assumptions about how this shoul

These items are mentioned in the [Google Doc](https://docs.google.com/document/d/1zAmXE4YNF0ut2lBagGlQXoMEEI8EoVpes7tc9gxeh0M/edit) I made after I reviewed Yash's proposal.

dixonwhitmire commented:
Ooops - hit the wrong button. 
@sjpadgett @bradymiller @yashrajbothra - still working on the FhirPatientService. Overall work remaining for a "reference implementation" for Yash's work.

- Support insert operation
- Support search including mapping FHIR search parameters to Open EMR fields
- Integrate FhirPatientRestController processing so that an OperationOutcome response is returned for validation errors.

I am unable to implement "getOne" and "update" since we don't have a means of associating a FHIR Resource ID to, in my case, an OpenEMR patient record. This is a larger decision we could use group feedback on. Some available options which have been discussed a bit:
- add a "fhir_resource_id" column to the patient_data table
- repurpose the patient_data.pid to be a uuid column capable of FHIR support. May not be trivial based on current "pid" usage
- add a new table for "fhir metadata" that  we can join onto.

bradymiller commented:
hi @dixonwhitmire , For the uuid issue, will every resource need to support this (or just Patient resource)? If we do create a fhir_metadata table, is there other important info that would also need to be included instead of just the uuid?

dixonwhitmire commented:
Hey @bradymiller ! Great question! 

All resources are required to support an "ID" field. If you would like to drill down into this a bit, here's a [link](https://www.hl7.org/fhir/resource.html) to the "base resource" definition which all FHIR resources extend and another [link](https://www.hl7.org/fhir/datatypes.html#id) to the spec for a valid id.

In regards to additional attributes, each FHIR resource contains a [meta](https://www.hl7.org/fhir/resource.html#Meta) that contains additional useful, but optional fields. Of these OpenEMR could return the "last updated" timestamp and a "1" for the version number. @sjpadgett @yashrajbothra and I discussed versioning a bit (in various PRs/threads) and concluded that versioning will not supported as we are focused on getting things working with the existing FHIR resources.

So that was a long answer - the short answer is that I don't believe OpenEMR needs to track additional FHIR attributes at this time 😄 I just wanted to provide links to the FHIR spec so that you could review, just in case.

dixonwhitmire commented:
And an update on PR progress - I am currently wiring up the search/getAll parameter translations and what not to get things working.

sjpadgett commented:
> Of these OpenEMR could return the "last updated" timestamp and a "1" for the version number

I think this is a great idea! However, we could take it a bit further and set up meta tracking for many other transactions, even standard API.
Fhir is not the only other interop we do that requires/begs for a uuid. CCDA, labs, procedures and various external support REST APIs come to mind. Also, i'm on a mission to track encounter history(thinking important for institutional settings).

We could continue this for implementation thoughts on this issue #2767 which I reopened for this purpose.

yashrajbothra commented:
> * add a new table for "fhir metadata" that  we can join onto.

Rather than adding a new table can we create [seperate database](https://github.com/openemr/openemr/issues/2767) which will help us with `_history` and keeping meta data ?

dixonwhitmire commented:
I've merged in the latest commits which include support for the uuid registry. I had a few test cases errors that I think I can attribute to "caching errors" as recreating the docker compose stack and enabling the APIs in the UI resolved the issues.

So, it's back to finishing up the PatientService!

dixonwhitmire commented:
@bradymiller @sjpadgett @yashrajbothra - Hi Folks I finally have the patient fhir and openemr apis integrated with the "standard api process" I added a few PRs ago. Additionally, I have updated the branch to support all of the UUID stuff Brady has been working on. 

This has been a rather long running branch. Please let me know if you would like me to squash or rebase and resubmit. Although that may be tricky as since I've been merging in "upstream" master along the way.

Thanks!

dixonwhitmire commented on .travis.yml:
```diff
@@ -280,6 +280,23 @@ language: php
       jobTests+="${jobTest}\n"
       echo "${jobTest}"
       echo "----------------------"
+    - &build_test_common |
```
For the UuidRegistry tests

dixonwhitmire commented on _rest_routes.inc.php:
```diff
@@ -355,13 +355,28 @@
         $data = (array) RestConfig::getPostData((file_get_contents("php://input")));
         return (new AuthRestController())->authenticate($data);
     },
+    "POST /fhir/Patient" => function () {
```
just grouping these together

dixonwhitmire commented on src/Common/Uuid/UuidRegistry.php:
```diff
@@ -120,4 +121,22 @@ public function tableNeedsUuidCreation()
         }
         return false;
     }
+
```
UUID conversion functions

dixonwhitmire commented on src/RestControllers/FHIR/FhirPatientRestController.php:
```diff
@@ -9,119 +9,109 @@
 use OpenEMR\FHIR\R4\FHIRResource\FHIRBundle\FHIRBundleEntry;
+use OpenEMR\Validators\ProcessingResult;
+
+require_once('_rest_config.php');
```
I found I had to do this to get the import for RestConfig (class in global namespace) to work.

dixonwhitmire commented on src/RestControllers/FHIR/FhirPatientRestController.php:
```diff
@@ -9,119 +9,109 @@
         $fhirValidate = $this->fhirValidate->validate($fhirJson);
         if (!empty($fhirValidate)) {
-            return RestControllerHelper::responseHandler($fhirValidate, null, 404);
+            return RestControllerHelper::responseHandler($fhirValidate, null, 400);
```
validation errors should return a 400 - bad client request

dixonwhitmire commented on src/RestControllers/FHIR/FhirPatientRestController.php:
```diff
@@ -9,119 +9,109 @@
-            return RestControllerHelper::responseHandler($fhirValidate, null, 404);
+            return RestControllerHelper::responseHandler($fhirValidate, null, 400);
         }
-        $data = $this->fhirPatientService->parsePatientResource($fhirJson);
```
all of the parsing, validating (most of  the validating) occurs in the service layer now, which makes the controller a bit simpler.

dixonwhitmire commented on src/RestControllers/RestControllerHelper.php:
```diff
@@ -38,7 +37,13 @@ public static function validationHandler($validationResult)
             http_response_code(400);
-            return $validationResult->getMessages();
+
+            if (property_exists($validationResult, 'getValidationMessages')) {
```
This may not be an update which "sticks" in the final FHIR implementation, but it allows the code to support "un-updated" endpoints and the patient endpoint.

dixonwhitmire commented on src/Services/BaseService.php:
```diff
@@ -106,6 +107,10 @@ protected function buildUpdateColumns($passed_in = array())
             if (in_array($key, array_column($this->autoIncrements, 'Field'))) {
                 continue;
             }
+            // exclude uuid columns
```
The uuid columns should be considered immutable and not available for updates.
If for some reason it could be updateable you would need to implement the uuid conversions for things to work correctly.

dixonwhitmire commented on src/Services/FHIR/FhirServiceBase.php:
```diff
@@ -0,0 +1,158 @@
+ * @copyright Copyright (c) 2020 Dixon Whitmire <dixonwh@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+abstract class FhirServiceBase
```
I added a base class for the FHIR services. Hopefully the docs are self explanatory.

dixonwhitmire commented on src/Services/FHIR/FhirServiceBase.php:
```diff
@@ -0,0 +1,158 @@
+
+    /**
+     * Updates an existing OpenEMR record.
+     * NOT YET IMPLEMENTED - requires updates to support associating a fhir resource id to an openemr record
```
oops I need to update the docstring here and below.

dixonwhitmire commented on src/Validators/PatientValidator.php:
```diff
@@ -17,6 +19,25 @@
+    public function isExistingUuid($uuid)
+    {
+        try {
+            $uuidLookup = UuidRegistry::uuidToBytes($uuid);
```
example of uuid conversion

dixonwhitmire commented:
rebased this PR and submitted as #3558
